### PR TITLE
Respect PreferParameterNullChecking user option

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes https://github.com/dotnet/roslyn/issues/59559

Relevant Discord discussion that lead to this PR:

image

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:36]: 💡 Consider rewriting this to minimize unnecessary/duplicate parameterization of the method:

```suggestion
        [InlineData(null)]
        [InlineData(true)]
        [InlineData(false)]
        public async Task TestSimpleReferenceType(bool? useParameterNullChecking)
```

**Review by sharwell — APPROVED** [review]

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (RikkiGibson) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:36]: My preference would be to extract the scenario where the option is explicitly disabled to its own `[Fact]`, and then treat this test as though `useParameterNullChecking` is a constant true. However, I don't feel strongly about any particular factoring of the test.

**Comment by RikkiGibson** [comment]
Please advise us if you plan to make any change to the test, and otherwise I think we can just merge.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (Youssef1313) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:36]: @sharwell @RikkiGibson I'm a bit confused as to which suggestion should I take. Both sounds reasonable and I'm okay either way.

Comment (RikkiGibson) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:36]: Let's go with https://github.com/dotnet/roslyn/pull/59561#discussion_r807190446 :)

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (RikkiGibson) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:33]: It looks like the case where the option is explicitly enabled was removed. Was that intended?

**Review by RikkiGibson — APPROVED** [review]

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: The downside of the approach mentioned by @RikkiGibson is we lose the ability to see that the two tests are operating on different input. At that point, it's not clear whether the different output is caused by a change in the option, or a change in the input. While this is rarely a problem in the pull request where the test is introduced, it's easy and common for an update to one test to fail to realize that the other is a matching test that needs to be updated at the same time. For this reason, it is strongly recommended for the test to be combined.

Comment (sharwell) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: @Youssef1313 I won't block this PR on the suggested change. More a note for the future to keep with the approach you were already using 😄

**Comment by sharwell** [comment]
📝 CLA bot seems to be down. Have reported this and will ping that check after it's fixed/online again to trigger the merge.

**Pull request merged**
